python/semantic_kernel/plugin_definition/kernel_plugin_collection.py

```diff
@@ -0,0 +1,249 @@
+        elif isinstance(plugins, Iterable):
+            # Process an iterable of plugins
+            plugins = self._process_plugins_iterable(plugins)
+        else:
+            raise ValueError("Invalid type for plugins")
+
+        super().__init__(plugins=plugins)
+
+    @staticmethod
+    def _process_plugins_iterable(plugins_input: Iterable[KernelPluginBase]) -> Dict[str, KernelPluginBase]:
+        plugins_dict = {}
+        for plugin in plugins_input:
+            if plugin is None:
+                raise ValueError("Plugin and plugin.name must not be None")
+            if plugin.name in plugins_dict:
+                raise ValueError(f"Duplicate plugin name detected: {plugin.name}")
+            plugins_dict[plugin.name] = plugin
+        return plugins_dict
+
+    def add(self, plugin: KernelPluginBase) -> None:
+        """
+        Add a single plugin to the collection
+
+        Args:
+            plugin (KernelPluginBase): The plugin to add to the collection.
+
+        Raises:
+            ValueError: If the plugin or plugin.name is None.
+        """
+        if plugin is None or plugin.name is None:
+            raise ValueError("Plugin and plugin.name must not be None")
+        if plugin.name in self.plugins:
+            raise ValueError(f"Plugin with name {plugin.name} already exists")
+        self.plugins[plugin.name] = plugin
+
+    def add_plugin_from_function(self, plugin_name: str, function: "KernelFunctionBase") -> None:
+        """
+        Add a function to a new plugin in the collection
+
+        Args:
+            plugin_name (str): The name of the plugin to create.
+            function (KernelFunctionBase): The function to add to the plugin.
+
+        Raises:
+            ValueError: If the function or plugin_name is None or invalid.
+        """
+        if not function or not plugin_name:
+            raise ValueError("Function and plugin_name must not be None or empty")
+        if plugin_name in self.plugins:
+            raise ValueError(f"Plugin with name {plugin_name} already exists")
+
+        plugin = DefaultKernelPlugin.from_function(plugin_name=plugin_name, function=function)
+        self.plugins[plugin_name] = plugin
+
+    def add_functions_to_plugin(self, functions: List["KernelFunctionBase"], plugin_name: str) -> None:
+        """
+        Add a function to a plugin in the collection
+
+        Args:
+            functions (List[KernelFunctionBase]): The function to add to the plugin.
+            plugin_name (str): The name of the plugin to add the function to.
+
+        Raises:
+            ValueError: If the function or plugin_name is None or invalid.
+            ValueError: if the function already exists in the plugin.
+        """
+        if not functions or not plugin_name:
+            raise ValueError("Functions and plugin_name must not be None or empty")
+        if plugin_name not in self.plugins:
+            self.plugins.add(DefaultKernelPlugin(name=plugin_name))
```

Comment:
can't you add the functions directly to this plugin creation?

Reply:
Technically yes, but I want the code below this to process each function to check for duplicate function names (which is not allowed).

Reply:
Talked offline. Update incoming.